### PR TITLE
Implement email sender service as part of a notification microservice

### PR DESCRIPTION
**Pull request by bogdanmilojevic**

### CODE REVIEW

**Review by bojanabogicevic — APPROVED** [review]

File: configuration-service/secrets.sh [hunk 1]
```diff
@@ -15,4 +15,5 @@ echo "Adding secrets..."
...
+vault kv put dev/notification-service spring.mail.username=milojevicbogdan01@gmail.com spring.mail.password='qeiq zshi feun rptb'
```
Comment (ilicstef94) [inline configuration-service/secrets.sh:18]: Should such things be pushed on repository?

File: configuration-service/src/main/resources/configs/notification-service.yml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+eureka:
+  client:
+    service-url:
+      default-zone: http://localhost:8761/eureka
```
Comment (ilicstef94) [inline configuration-service/src/main/resources/configs/notification-service.yml:4]: Values that change depending on the environment (in this case: host, port) should be moved to be in the .env file.

File: notification-service/src/main/java/com/microcommerce/notificationservice/listener/AccountCreatedEventListener.java [hunk 3]
```diff
@@ -14,19 +15,10 @@
 @RequiredArgsConstructor
 public class AccountCreatedEventListener {
 
-    private final EmailSenderService emailSenderService;
+    private final AccountCreatedEventHandlerService accountCreatedEventHandler;
 
-    @KafkaListener(topics = "${kafka.topics.account-topic}", groupId = "group-id-2", containerFactory = "accountCreatedListenerContainerFactory")
+    @KafkaListener(topics = "${kafka.topics.account-topic.name}", groupId = "${kafka.topics.account-topic.group-id}", containerFactory = "accountCreatedListenerContainerFactory")
```
Comment (ilicstef94) [inline notification-service/src/main/java/com/microcommerce/notificationservice/listener/AccountCreatedEventListener.java:20]: Is there any way of message acknowledgment in the situation when the mail is actually delivered. How to handle the situation when the processing of the message from the topic is not successful. If necessary, please make an investigation ticket and propose the best solution.

File: notification-service/src/main/java/com/microcommerce/notificationservice/service/AccountCreatedEventHandlerService.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import org.thymeleaf.TemplateEngine;
+import org.thymeleaf.context.Context;
+
+@Service
+@Slf4j
+@RequiredArgsConstructor
+public class AccountCreatedEventHandlerService {
+    private static final String ACCOUNT_CONFIRMATION_EMAIL_TEMPLATE = "account_confirmation_template";
+    private static final String UTF_8_ENCODING = "UTF-8";
+    private static final String ACCOUNT_CONFIRMATION_URL = "http://localhost:8080/auth/confirm?token=";
```
Comment (ilicstef94) [inline notification-service/src/main/java/com/microcommerce/notificationservice/service/AccountCreatedEventHandlerService.java:20]: Values that change depending on the environment (in this case: host, port) should be moved to be in the .env file.

File: notification-service/src/main/resources/application.yml [hunk 5]
```diff
@@ -25,19 +18,20 @@ spring:
             enable: true
             required: true
     verify:
-      host: ${VERIFY_EMAIL_HOST}
+      host: http://localhost:587
```
Comment (ilicstef94) [inline notification-service/src/main/resources/application.yml:21]: Values that change depending on the environment (in this case: host, port) should be moved to be in the .env file.

File: notification-service/src/main/resources/application.yml [hunk 6]
```diff
@@ -25,19 +18,20 @@ spring:
...
+    notification-topic:
+      name: notification-topic
+      group-id: group-id
+    account-topic:
+      name: account-topic
+      group-id: group-id-2
 
-server:
-  port: 8083
+account-confirmation-url: "http://localhost:8080/auth/confirm?token="
```
Comment (ilicstef94) [inline notification-service/src/main/resources/application.yml:34]: Values that change depending on the environment (in this case: host, port) should be moved to be in the .env file.

File: notification-service/src/main/resources/bootstrap.yml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+spring:
+  cloud:
+    config:
+      uri: http://localhost:8888/config
```
Comment (ilicstef94) [inline notification-service/src/main/resources/bootstrap.yml:4]: Values that change depending on the environment (in this case: host, port) should be moved to be in the .env file.

File: user-service/src/main/java/com/microcommerce/userservice/data/entity/UserRole.java [hunk 8]
```diff
@@ -10,10 +10,12 @@
 @Getter
 @Setter
 @NoArgsConstructor
+@Builder
+@AllArgsConstructor
 public class UserRole {
 
     @Id
-    @GeneratedValue(strategy = GenerationType.UUID)
+    @GeneratedValue(strategy = GenerationType.SEQUENCE)
```
Comment (ilicstef94) [inline user-service/src/main/java/com/microcommerce/userservice/data/entity/UserRole.java:18]: Why sequence for this table?

**Review by ilicstef94 — CHANGES_REQUESTED** [review]

**Comment by ilicstef94** [comment]
Note that wherever there are values that change depending on the environment, they need to be moved to the .env file so during the deployment you can easily proper values for specific environment.

File: user-service/src/main/java/com/microcommerce/userservice/data/entity/UserRole.java (hunk 8 above)
Comment (bogdanmilojevic) [reply user-service/src/main/java/com/microcommerce/userservice/data/entity/UserRole.java:18]: Type of the id here is "Long" so the strategy can't be GenerationType.UUID, that change broke the code when it was introduced.

File: notification-service/src/main/resources/application.yml (hunk 6 above)
Comment (bogdanmilojevic) [reply notification-service/src/main/resources/application.yml]: Done.

File: notification-service/src/main/resources/bootstrap.yml (hunk 7 above)
Comment (bogdanmilojevic) [reply notification-service/src/main/resources/bootstrap.yml]: Done.

File: notification-service/src/main/resources/application.yml (hunk 5 above)
Comment (bogdanmilojevic) [reply notification-service/src/main/resources/application.yml]: Done.

File: notification-service/src/main/java/com/microcommerce/notificationservice/service/AccountCreatedEventHandlerService.java (hunk 4 above)
Comment (bogdanmilojevic) [reply notification-service/src/main/java/com/microcommerce/notificationservice/service/AccountCreatedEventHandlerService.java]: Done.

File: configuration-service/secrets.sh (hunk 1 above)
Comment (bogdanmilojevic) [reply configuration-service/secrets.sh]: Removed

File: notification-service/src/main/java/com/microcommerce/notificationservice/listener/AccountCreatedEventListener.java (hunk 3 above)
Comment (bogdanmilojevic) [reply notification-service/src/main/java/com/microcommerce/notificationservice/listener/AccountCreatedEventListener.java:20]: There is a way to manually send acknowledgement to kafka broker using Acknowledgement parameter in listener methods, ticket is created: https://github.com/rbtree/micro-commerce/issues/26

**Review by ilicstef94 — APPROVED** [review]

**Pull request merged**